pkg/containerruntime/containerd_config.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2021 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package containerruntime
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/BurntSushi/toml"
+
+	"k8c.io/kubeone/pkg/apis/kubeone"
+)
+
+type containerdConfig struct {
+	Version int                    `toml:"version"`
+	Metrics *containerdMetrics     `toml:"metrics"`
+	Plugins map[string]interface{} `toml:"plugins"`
+}
+
+type containerdMetrics struct {
+	Address string `toml:"address"`
+}
+
+type containerdCRIPlugin struct {
+	Containerd *containerdCRISettings `toml:"containerd"`
+	Registry   *containerdCRIRegistry `toml:"registry"`
+}
+
+type containerdCRISettings struct {
+	Runtimes map[string]containerdCRIRuntime `toml:"runtimes"`
+}
+
+type containerdCRIRuntime struct {
+	RuntimeType string      `toml:"runtime_type"`
+	Options     interface{} `toml:"options"`
+}
+
+type containerdCRIRuncOptions struct {
+	SystemdCgroup bool
+}
+
+type containerdCRIRegistry struct {
+	Mirrors map[string]containerdMirror `toml:"mirrors"`
+}
+
+type containerdMirror struct {
+	Endpoint []string `toml:"endpoint"`
+}
+
+func marshalContainerdConfig(cluster *kubeone.KubeOneCluster) (string, error) {
+	criPlugin := containerdCRIPlugin{
+		Containerd: &containerdCRISettings{
+			Runtimes: map[string]containerdCRIRuntime{
+				"runc": {
+					RuntimeType: "io.containerd.runc.v2",
+					Options: containerdCRIRuncOptions{
+						SystemdCgroup: true,
+					},
+				},
+			},
+		},
+		Registry: &containerdCRIRegistry{
+			Mirrors: map[string]containerdMirror{
+				"docker.io": {
+					Endpoint: []string{"https://registry-1.docker.io"},
+				},
+			},
+		},
+	}
+
+	insecureRegistry := cluster.RegistryConfiguration.InsecureRegistryAddress()
+	if insecureRegistry != "" {
+		criPlugin.Registry.Mirrors[insecureRegistry] = containerdMirror{
+			Endpoint: []string{fmt.Sprintf("http://%s", insecureRegistry)},
+		}
+	}
+
+	if reg := cluster.RegistryConfiguration; reg != nil {
+		if len(reg.Mirrors) > 0 {
+			criPlugin.Registry.Mirrors["docker.io"] = containerdMirror{
+				Endpoint: reg.Mirrors,
+			}
```

Comment:
Why don't we support mirrors for registries other than `docker.io`? I assume that Docker doesn't support that, but if containerd does, it could be useful.

Reply:
Indeed, the docker is kinda anchor for such features that we can't supported in both. And to overcome this limitation I wanted to introduce different settings for different runtimes inside new v1beta2 API. So that' something to do in the future.

Reply:
I think that we can easily overcome this limitation in the current API version as well. Let's say we have a struct such as:

```go
type RegistryMirror struct {
    Registry string
    Mirror string
}

...
  Mirrors []RegistryMirror
```

You can easily extend the validation to reject the config if the container runtime is Docker and there's a registry mirror that's not `docker.io`.

Reply:
Point is, I don’t want to change v1beta1 even more, given that we are about to have v1beta2

Reply:
But we are already changing the v1beta1 API. I'm a little bit conflicted on providing two different versions of a feature in different API versions, but in the same release.

It would make total sense if we introduced this in some earlier release and then decided to improve it in the new API version. In that case, leaving v1beta1 API unchanged would probably be the best. However, in this concrete case, we're introducing a new feature in two APIs at the same time (v1beta1 and v1beta2), but we don't provide the same feature set. That doesn't make much sense to me and it can be confusing to users too. Such cases also make migration to v1beta2 more complicated.

I would prefer either leaving this only for v1beta2 or doing what I proposed above (I think I would prefer this approach).

Reply:
Leaving this for v1beta2 is totally OK and I'd prefer that. In v1beta2 we should delete RegistryConfiguration and deconstruct to two parts inside `ContainerRuntimeContainerd` and `ContainerRuntimeDocker`, each with own fields and possibilities.

Reply:
However I don't want to waste non-API related changes from this PR (refactoring of how whole container-runtime config generation works) and will repackage it in another PR.

Reply:
I'm totally fine with leaving this for v1beta2. :+1: 

We discussed it already earlier -- you can just update the internal API version for now, and leave v1beta1 untouched. You don't even need to change the API much, we can always change the internal API.